### ISSUE TITLE
[BUG] Game state transition causing mutex error and crashing

### ISSUE THREAD

**Issue by tENCoRtO** [amethyst/amethyst#2142]
## Description

I ran into problem setting up a flow between different game states where a mutex is failing to lock, and once or three times I got a segfault. I'm not sure what I'm doing wrong, and I can't find any documentation or examples showing this kind of game state flow. I've set up a minimum example of the failure **[here](https://github.com/tENCoRtO/transition_fail)**.

I'm pretty sure I'm handling states, assets, entities, or components incorrectly as they seem to persist in some form after being released--I know entities for example are recycled. Is there a particular way I need to be releasing resources? Just spit-balling here, I really don't know what I'm doing wrong D:

## Reproduction Steps

From the [example readme](https://github.com/tENCoRtO/transition_fail/blob/master/README.md):

Trigger a mutex error or occasional segfault by:

```
RUST_BACKTRACE=1 AMETHYST_LOG_LEVEL_FILTER=trace cargo run
```

|**Step**|**Stack Op**|**Stack After**|**It...**|
|--------|------------|---------------|----------|
|1. Launch the game|push|`[MainMenuState]`|Succeeds|
|2. Click "Start Game"|push|`[MainMenuState, MainGameState]`|Succeeds|
|3. Press "Esc"|push|`[MainMenuState, MainGameState, PauseMenuState]`|Succeeds|
|4. Click "Exit to Menu"|replace|`[MainMenuState]`|Succeeds|
|5. Click "Start Game"|push|`[MainMenuState, MainGameState]`|Crashes|

## What You Expected to Happen

I expect a clean transition from the `MainMenuState` to `MainGameState`.

## What Actually Happened

```
libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
```

## Screenshots and Logs

[TRACE-level log attached](https://github.com/amethyst/amethyst/files/4231689/log.txt)

[Log with vulkan validation layers](https://github.com/amethyst/amethyst/files/4238106/full_log.txt)

## Your Environment

1. Operating System: **macos**
2. Operating System Version: **10.15.3 (catalina)**
3. Version of Amethyst: **0.14.0**

And just in case it matters! **Rust 1.41.0**

**Comment by ishitatsuyuki** [comment]
You're using MoltenVK without validation layers, and that's not good for troubleshooting. The exception is likely triggered within MoltenVK (due to improper Vulkan API calls), which you need validation layer in order to get a meaningful error message.

**Comment by tENCoRtO** [comment]
Awesome to know about validation layers! I found solid instructions to enable them on macos on the [ash crate's readme](https://github.com/MaikKlein/ash#macos).

The logs are pretty noisy, looks like there's a few different things going on that vulkan doesn't like. The big blocks of `VALIDATION [VUID-VkMappedMemoryRange-size-01390 (0)]` errors happen both in my code and in the amethyst examples, so I assume those aren't related to this specific issue.

There's a bunch of errors similar to this one stand out though when the game crashes:
```
[ERROR][gfx_backend_vulkan]
VALIDATION [UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotUpdated (0)] : VkDescriptorSet 0xe400000000e4[] bound as set #1 encountered the following validation error at vkCmdDraw() time: Descriptor in binding #1 index 0 is using imageView VkImageView 0x8c000000008c[] that is invalid or has been destroyed.
object info: (type: DESCRIPTOR_SET, hndl: 250688651133156)
```

I've attached a [truncated log](https://github.com/amethyst/amethyst/files/4238107/truncated_log.txt) of errors that occur after the final button press, and also the [complete log](https://github.com/amethyst/amethyst/files/4238106/full_log.txt) for reference.
